File: src/main/kotlin/samlang/checker/TypeValidator.kt
package samlang.checker

import samlang.ast.Range
import samlang.ast.Type
import samlang.ast.Type.FunctionType
import samlang.ast.Type.IdentifierType
import samlang.ast.Type.PrimitiveType
import samlang.ast.Type.TupleType
import samlang.ast.Type.UndecidedType
import samlang.ast.TypeVisitor

internal fun Type.validate(context: TypeCheckingContext, errorRange: Range): Type =
    accept(visitor = TypeValidator(errorRange = errorRange), context = context)

/**
 * A validator for type to check whether every identifier type is well defined.
 */
private class TypeValidator(private val errorRange: Range) : TypeVisitor<TypeCheckingContext, Type> {

    override fun visit(type: PrimitiveType, context: TypeCheckingContext): Type = type

    override fun visit(type: IdentifierType, context: TypeCheckingContext): Type {
        val (name, typeArguments) = type
        context.checkIfIdentifierTypeIsWellDefined(
            name = name,
            typeArgLength = typeArguments?.size ?: 0,
            errorRange = errorRange
        )
        return type.copy(typeArguments = typeArguments?.map { it.accept(visitor = this, context = context) })
    }

    override fun visit(type: TupleType, context: TypeCheckingContext): Type =
        type.copy(mappings = type.mappings.map { it.accept(visitor = this, context = context) })

    override fun visit(type: FunctionType, context: TypeCheckingContext): Type =
        type.copy(
            argumentTypes = type.argumentTypes.map { it.accept(visitor = this, context = context) },
            returnType = type.returnType.accept(visitor = this, context = context)
        )

    override fun visit(type: UndecidedType, context: TypeCheckingContext): Type = type
}

File: samlang-core/checker/type-validator.ts
import type { Type, Range, ModuleReference } from '../ast/common-nodes';
import type { ModuleErrorCollector } from '../errors';
import { checkNotNull } from '../utils';

export interface IdentifierTypeValidator {
  /**
   * Given the `name` of the identifier and the number of applied arguments, check the context to see whether it
   * matches any type definition in scope.
   */
  identifierTypeIsWellDefined(
    moduleReference: ModuleReference,
    name: string,
    typeArgumentLength: number
  ): boolean;
}

/**
 * @returns an invalidid string identifier, null if there is none.
 */
export function findInvalidTypeIdentifier_EXPOSED_FOR_TESTING(
  type: Type,
  identifierTypeValidator: IdentifierTypeValidator
): string | null {
  switch (type.type) {
    case 'PrimitiveType':
      return null;
    case 'IdentifierType':
      if (
        !identifierTypeValidator.identifierTypeIsWellDefined(
          type.moduleReference,
          type.identifier,
          type.typeArguments.length
        )
      ) {
        return type.identifier;
      }
      return findInvalidTypeIdentifierForList(type.typeArguments, identifierTypeValidator);
    case 'TupleType':
      return findInvalidTypeIdentifierForList(type.mappings, identifierTypeValidator);
    case 'FunctionType':
      return (
        findInvalidTypeIdentifier_EXPOSED_FOR_TESTING(type.returnType, identifierTypeValidator) ??
        findInvalidTypeIdentifierForList(type.argumentTypes, identifierTypeValidator)
      );
    case 'UndecidedType':
      return null;
  }
}

function findInvalidTypeIdentifierForList(
  types: readonly Type[],
  identifierTypeValidator: IdentifierTypeValidator
): string | null {
  for (let i = 0; i < types.length; i += 1) {
    const invalidName = findInvalidTypeIdentifier_EXPOSED_FOR_TESTING(
      checkNotNull(types[i]),
      identifierTypeValidator
    );
    if (invalidName != null) return invalidName;
  }
  return null;
}

export function validateType(
  type: Type,
  identifierTypeValidator: IdentifierTypeValidator,
  errorCollector: ModuleErrorCollector,
  errorRange: Range
): boolean {
  const badIdentifier = findInvalidTypeIdentifier_EXPOSED_FOR_TESTING(
    type,
    identifierTypeValidator
  );
  if (badIdentifier == null) {
    return true;
  }
  errorCollector.reportNotWellDefinedIdentifierError(errorRange, badIdentifier);
  return false;
}
